File: analysis/import/SCMLogParser/src/main/java/de/maibornwolff/codecharta/model/input/VersionControlledFile.java
package de.maibornwolff.codecharta.model.input;

import de.maibornwolff.codecharta.model.input.metrics.CommitMetric;
import de.maibornwolff.codecharta.model.input.metrics.Metric;
import de.maibornwolff.codecharta.model.input.metrics.MetricsFactory;
import de.maibornwolff.codecharta.model.input.metrics.ModificationMetric;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class VersionControlledFile {

    private final String filename;
    private final Set<String> authors = new HashSet<>();
    private final List<ModificationMetric> modificationMetrics;
    private final List<CommitMetric> commitMetrics;

    public VersionControlledFile(String filename, MetricsFactory metricsFactory) {
        this(filename, metricsFactory.createModificationMetrics(), metricsFactory.createCommitMetrics());
    }

    VersionControlledFile(
            String filename,
            List<ModificationMetric> modificationMetrics,
            List<CommitMetric> commitMetrics
    ) {
        this.filename = filename;
        this.modificationMetrics = modificationMetrics;
        this.commitMetrics = commitMetrics;
    }

    public void registerCommit(Commit commit) {
        Optional<Modification> modification = commit.getModification(filename);
        if (modification.isPresent()) {
            modificationMetrics.forEach(m -> m.registerModification(modification.get()));
            commitMetrics.forEach(m -> m.registerCommit(commit));
            authors.add(commit.getAuthor());
        }
    }

    public List<Metric> getMetrics() {
        return Stream.of(modificationMetrics, commitMetrics)
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
    }

    public Map<String, Number> getMetricsMap() {
        return getMetrics().stream().collect(Collectors.toMap(Metric::metricName, Metric::value));
    }

    public String getFilename() {
        return filename;
    }

    public Set<String> getAuthors() {
        return authors;
    }

    @Override
    public String toString() {
        return filename + " with metrics " + getMetricsMap();
    }

    public int getMetricValue(String metricName) {
        return getMetrics().stream()
                .filter(m -> m.metricName().equals(metricName)).findAny()
                .orElseThrow(() -> new RuntimeException("metric " + metricName + " not found."))
                .value().intValue();
    }
}

File: analysis/import/SCMLogParserV2/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparserv2/input/VersionControlledFile.kt
package de.maibornwolff.codecharta.importer.scmlogparserv2.input

import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.Metric
import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.MetricsFactory
import de.maibornwolff.codecharta.model.Edge
import java.util.Arrays

class VersionControlledFile internal constructor(
    filename: String,
    private var metrics: List<Metric>
) {

    val authors = mutableSetOf<String>()
    private val renames = mutableSetOf<String>()

    // the current filename in a specific revision, might change in history
    var filename: String

    private var deleted: Boolean = false
    private var mutated: Boolean = false

    val metricsMap: Map<String, Number>
        get() = metrics.associateBy({ it.metricName() }, { it.value() })

    constructor(filename: String, metricsFactory: MetricsFactory) : this(filename, metricsFactory.createMetrics())

    internal constructor(filename: String, vararg metrics: Metric) : this(filename, Arrays.asList<Metric>(*metrics))

    init {
        this.filename = filename
        addRename(filename.substringBefore("_\\0_"))
    }

    /**
     * registers commits in chronological order
     */
    fun registerCommit(commit: Commit, mod: Modification) {

        flagCommitAsMutatedIfNeeded(commit, mod)

        metrics.forEach {
            it.registerCommit(commit)
            it.registerModification(mod)
        }
        authors.add(commit.author)
    }

    fun containsRename(rename: String): Boolean {
        return renames.contains(rename)
    }

    fun addRename(rename: String) {
        renames.add(rename)
    }

    override fun toString(): String {
        return "$filename with metrics $metricsMap"
    }

    fun getEdgeList(): List<Edge> {
        val edgeMap = mutableMapOf<String, Edge>()
        metrics.flatMap { it.getEdges() }
            .forEach { edge ->
                val resolvedEdge = edgeMap[edge.toNodeName]
                if (resolvedEdge != null) {
                    edge.attributes.toMutableMap().putAll(edge.attributes)
                } else {
                    edgeMap[edge.toNodeName] = edge
                }
            }
        return edgeMap.values.toList()
    }

    private fun flagCommitAsMutatedIfNeeded(commit: Commit, mod: Modification) {
        if (this.isDeleted() && (mod.isTypeModify() || mod.isTypeRename())) {
            this.mutate()
        } else if (!commit.isMergeCommit() && !this.isDeleted() && mod.isTypeAdd() && !mod.isInitialAdd()) {
            this.mutate()
        }
    }

    fun getMetricValue(metricName: String): Number {
        return metricsMap[metricName] ?: error("No element found")
    }

    fun removeMetricsToFreeMemory() {
        metrics = listOf()
    }

    fun markDeleted() {
        deleted = true
    }

    fun unmarkDeleted() {
        deleted = false
    }

    fun isDeleted(): Boolean {
        return deleted
    }

    fun mutate() {
        mutated = true
    }

    fun resetMutation() {
        mutated = false
    }

    fun isMutated(): Boolean {
        return mutated
    }
}
